mackerel/api.go

```diff
@@ -68,243 +48,30 @@ func infoError(msg string) *InfoError {
 
 // NewAPI creates a new instance of API.
 func NewAPI(rawurl string, apiKey string, verbose bool) (*API, error) {
-	u, err := url.Parse(rawurl)
+	c, err := mkr.NewClientWithOptions(apiKey, rawurl, verbose)
 	if err != nil {
 		return nil, err
 	}
-	return &API{BaseURL: u, APIKey: apiKey, Verbose: verbose}, nil
-}
-
-func (api *API) urlFor(path string, query string) *url.URL {
-	newURL, _ := url.Parse(api.BaseURL.String())
-	newURL.Path = path
-	newURL.RawQuery = query
-	return newURL
-}
-
-func (api *API) getUA() string {
-	if api.UA != "" {
-		return api.UA
-	}
-	return "mackerel-agent/0.0.0"
+	c.AdditionalHeaders = make(http.Header)
+	// TODO(lufia): should we set a timeout explicitly?
+	//c.HTTPClient.Timeout = apiRequestTimeout
+	return &API{Client: c}, nil
 }
 
 var apiRequestTimeout = 30 * time.Second
 
-func (api *API) do(req *http.Request) (resp *http.Response, err error) {
-	if api.DefaultHeaders != nil {
-		for k, vs := range api.DefaultHeaders {
-			for _, v := range vs {
-				req.Header.Add(k, v)
-			}
-		}
-	}
-	req.Header.Add("X-Api-Key", api.APIKey)
-	req.Header.Set("User-Agent", api.getUA())
-
-	if api.Verbose {
-		dump, err := httputil.DumpRequest(req, true)
-		if err == nil {
-			logger.Tracef("%s", dump)
```

Comment:
Logger prefix `<api>` for log in verbose mode will be lost but maybe ok.

Reply:
I've fixed at edde7d6